lib/float/common.sail

```diff
@@ -0,0 +1,84 @@
+/*                                                                          */
+/* Redistribution and use in source and binary forms, with or without       */
+/* modification, are permitted provided that the following conditions are   */
+/* met:                                                                     */
+/*                                                                          */
+/* 1. Redistributions of source code must retain the above copyright        */
+/*    notice, this list of conditions and the following disclaimer.         */
+/* 2. Redistributions in binary form must reproduce the above copyright     */
+/*    notice, this list of conditions and the following disclaimer in the   */
+/*    documentation and/or other materials provided with the distribution.  */
+/*                                                                          */
+/* THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS      */
+/* "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT        */
+/* LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A  */
+/* PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT       */
+/* HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,   */
+/* SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED */
+/* TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR   */
+/* PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF   */
+/* LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING     */
+/* NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS       */
+/* SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.             */
+/*==========================================================================*/
+
+$ifndef _FLOAT_COMMON
+$define _FLOAT_COMMON
+
+infix 2 <<
+
+/* Floating point types definition */
+type fp_exception_flags = bits(5) /* Floating-point exception flags. */
+type fp_bits = { 'n, 'n in {16, 32, 64}. bits('n) } /* Floating-point in bits. */
+type fp_bits_x2 = { 'n, 'n in {16, 32, 64}. (bits('n), bits('n)) } /* Floating-point in bits. */
+
+/* Floating point constants */
+let fp_eflag_invalid : fp_exception_flags = 0b00001
+
+/* The val func declarations */
+val zeros : forall 'n, 'n >= 0. implicit('n) -> bits('n)
+val ones  : forall 'n, 'n >= 0. implicit('n) -> bits('n)
+
+/* The operator declarations */
+val operator << = pure "shiftl" : forall 'n. (bits('n), int) -> bits('n)
+
+/* The val func implementations */
+val      float_decompose : forall 'n, 'n in { 16, 32, 64 }.
+  bits('n) -> (
+    bits(1),
+    bits(if   'n == 16
+         then 5
+         else (if   'n == 32
+               then 8
+               else 11)
+    ),
+    bits(if   'n == 16
+         then 10
+         else (if   'n == 32
+               then 23
+               else 52)
+    ),
+  )
+function float_decompose(op) = {
```

Comment:
Would it make sense to use a struct with named fields here rather than a plain tuple?

Reply:
Sure thing, let me update it.